### PR TITLE
Check minimum kernel memory limit to be 4M

### PR DESCRIPTION
**Pull request by hqhq**
Fixes: #18405

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

### CODE REVIEW

**Comment by runcom** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐶

**Comment by thaJeztah** [comment]
I think we should mention this in the documentation; https://github.com/docker/docker/search?utf8=✓&q=%22--kernel-memory%22 (perhaps in `docs/reference/run.md`)

But you may want to wait until https://github.com/docker/docker/pull/18432 is merged, because that introduces quite some changes in that document :innocent:

**Comment by thaJeztah** [comment]
Also; this affects both the "cli" and the "api"? If so, we should add the right "impact" labels, and add a note to the API change log; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v122-api-changes

File: daemon/daemon_unix.go [hunk 1]
```diff
@@ -207,6 +207,9 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
 		logrus.Warnf("Your kernel does not support kernel memory limit capabilities. Limitation discarded.")
 		hostConfig.KernelMemory = 0
 	}
+	if hostConfig.KernelMemory > 0 && hostConfig.KernelMemory < 4194304 {
```
Comment (cpuguy83) [inline daemon/daemon_unix.go]: This should be a const and also used here: https://github.com/hqhq/docker/blob/hq_limit_kernel_memory/daemon/daemon_unix.go#L166

Comment (calavera) [inline daemon/daemon_unix.go]: @hqhq do you mind moving this to a constant? It should be good to go then.

**Comment by hqhq** [comment]
@thaJeztah @cpuguy83 @calavera Updated.

File: integration-cli/docker_cli_run_unix_test.go [hunk 2]
```diff
@@ -169,10 +169,19 @@ func (s *DockerSuite) TestRunWithKernelMemory(c *check.C) {
 	out, err := inspectField("test1", "HostConfig.KernelMemory")
 	c.Assert(err, check.IsNil)
 	c.Assert(out, check.Equals, "52428800")
+}
+
+func (s *DockerSuite) TestRunWithInvalidKernelMemory(c *check.C) {
```
Comment (thaJeztah) [inline integration-cli/docker_cli_run_unix_test.go]: not for this PR, but I noticed we don't have this test for `--memory` (or I overlooked it)

**Comment by thaJeztah** [comment]
Thanks @hqhq, docs LGTM (documentation CI failures are not related)

**Comment by thaJeztah** [comment]
ping @SvenDowideit @moxiegirl

File: integration-cli/docker_cli_run_unix_test.go (hunk 2 above)
Comment (cpuguy83) [inline integration-cli/docker_cli_run_unix_test.go]: It exists

**Comment by cpuguy83** [comment]
LGTM

**Comment by vdemeester** [comment]
Hum I guess we can move it to docs-review right ?

**Comment by thaJeztah** [comment]
LOL, thought it already was. I think this is okay to merge; docs changes are minimal

**Comment by thaJeztah** [comment]
Added impact labels, in case people are currently expecting values `< 4M` to be accepted

File: integration-cli/docker_cli_run_unix_test.go (hunk 2 above)
Comment (hqhq) [inline integration-cli/docker_cli_run_unix_test.go]: @thaJeztah It's `TestRunWithTooSmallMemoryLimit` ;)

**Pull request merged**